sdks/python/apache_beam/typehints/decorators.py

```diff
@@ -193,7 +200,7 @@ def __repr__(self):
         self.input_types, self.output_types)
 
 
-class WithTypeHints(object):
+class WithTypeHints(Generic[InT, OutT]):
```

Comment:
I'm afraid that if we add a second typing system here will create confusion. If I write a PTransform, do I specify an input type to Generic[] or do I use a @with_input_type decorator? Which one is enforced?

Reply:
True.  I think the somewhat unsatisfactory answer is that for the time being you need both: one for runtime checking and the other for static checking, until such a time as they can become the same.  I think trying to do that all in one PR is going to be too much.

off the top of my head, the order this should probably be done is:

1. support runtime type hints using `typing` module instead of `typehints`
2. add static type hints to the beam code and begin enforcing it using mypy (this PR)
3. support static validation of user pipelines (mypy plugin, etc)
4. support runtime validations based on `typing` annotations

There are a lot of "ifs" surrounding step 4.  We may need to get to python3-only first to avoid the pitfalls of type comments.  We may find that step 3 makes it less important.

Reply:
(1) is already mostly there, for the user-facing API at least. How much of (2) is required for (3) and how invasive it is seems to still be an open question, which this PR is giving good insight to. (Long term, it'd also be nice to have (2) for development as well.) Given that mypy can't even do `beam.Map(lambda x: (x, 1))` I think we'll have to rely a lot on (4). 

All will be much nicer once we can have python-3 style annotations in the code itself. We can't do that for our own codebase until Python 2 is LTS only, but user code could start using it much sooner.

Reply:
> How much of (2) is required for (3) and how invasive it is seems to still be an open question, which this PR is giving good insight to. 

agreed.

>  Given that mypy can't even do beam.Map(lambda x: (x, 1)) I think we'll have to rely a lot on (4).

I think people who want to use static typing for their pipelines will have to stop using lambdas.  So there will definitely be a split between those who are willing to make that adjustment and those who want to continue relying on the current interface for dynamic typing. 

> All will be much nicer once we can have python-3 style annotations in the code itself. We can't do that for our own codebase until Python 2 is LTS only, but user code could start using it much sooner.

When we get to that stage, we can use this tool to convert comments to annotations:  https://github.com/ilevkivskyi/com2ann